Sources/NIOPerformanceTester/main.swift

```diff
@@ -0,0 +1,505 @@
+        buffer.write(staticString: htmlASCIIOnlyStaticString)
+        buffer.clear()
+    }
+    return buffer.readableBytes
+}
+
+measureAndPrint(desc: "bytebuffer_write_http_response_some_nonascii_as_string") {
+    var buffer = ByteBufferAllocator().buffer(capacity: 16 * 1024)
+    for _ in 0..<20_000 {
+        writeExampleHTTPResponseAsString(buffer: &buffer)
+        buffer.write(string: htmlMostlyASCII)
+        buffer.clear()
+    }
+    return buffer.readableBytes
+}
+
+measureAndPrint(desc: "bytebuffer_write_http_response_some_nonascii_as_staticstring") {
+    var buffer = ByteBufferAllocator().buffer(capacity: 16 * 1024)
+    for _ in 0..<20_000 {
+        writeExampleHTTPResponseAsStaticString(buffer: &buffer)
+        buffer.write(staticString: htmlMostlyASCIIStaticString)
+        buffer.clear()
+    }
+    return buffer.readableBytes
+}
+
+try measureAndPrint(desc: "no-net_http1_10k_reqs_1_conn") {
+    final class MeasuringHandler: ChannelDuplexHandler {
+        typealias InboundIn = Never
+        typealias InboundOut = ByteBuffer
+        typealias OutboundIn = ByteBuffer
+
+        private var requestBuffer: ByteBuffer!
+        private var expectedResponseBuffer: ByteBuffer?
+        private var remainingNumberOfRequests: Int
+
+        private let completionHandler: (Int) -> Void
+        private let numberOfRequests: Int
+
+        init(numberOfRequests: Int, completionHandler: @escaping (Int) -> Void) {
+            self.completionHandler = completionHandler
+            self.numberOfRequests = numberOfRequests
+            self.remainingNumberOfRequests = numberOfRequests
+        }
+
+        func handlerAdded(ctx: ChannelHandlerContext) {
+            self.requestBuffer = ctx.channel.allocator.buffer(capacity: 512)
+            self.requestBuffer.write(string: """
+                                             GET /perf-test-2 HTTP/1.1\r
+                                             Host: example.com\r
+                                             X-Some-Header-1: foo\r
+                                             X-Some-Header-2: foo\r
+                                             X-Some-Header-3: foo\r
+                                             X-Some-Header-4: foo\r
+                                             X-Some-Header-5: foo\r
+                                             X-Some-Header-6: foo\r
+                                             X-Some-Header-7: foo\r
+                                             X-Some-Header-8: foo\r\n\r\n
+                                             """)
+        }
+
+        func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+            var buf = self.unwrapOutboundIn(data)
+            if self.expectedResponseBuffer == nil {
+                self.expectedResponseBuffer = buf
+            }
+            precondition(buf == self.expectedResponseBuffer, "got \(buf.readString(length: buf.readableBytes)!)")
```

Comment:
Should this be an `assert`?

Reply:
@normanmaurer no because we want to run it in release mode (as the tester is basically only ever run in release mode)